src/api_midlevel.jl

```diff
@@ -0,0 +1,47 @@
+# This file defines midlevel api wrappers. We include name normalization for methods that are
+# applicable to different hdf5 api-layers. We still try to adhere close proximity to the underlying
+# method name in the hdf5-library.
+
+"""
+    HDF5.set_extent_dims(dset::HDF5.Dataset, new_dims::Dims)
+
+Change the current dimensions of a dataset to `new_dims`, limited by
+`max_dims = get_simple_extent_dims(dset)[2]`. Reduction is possible and leads to loss of truncated data.
+"""
+function set_extent_dims(dset::Dataset, size::Dims)
+    checkvalid(dset)
+    h5d_set_extent(dset, hsize_t[reverse(size)...])
+end
+
+"""
+    HDF5.set_extent_dims(dset::HDF5.Dataspace, new_dims::Dims, max_dims::Union{Dims,Nothing} = nothing)
+
+Change the current dimensions of a dataspace `dset` to `new_dims`, limited by
+`max_dims = get_simple_extent_dims(dset)[2]`. Reduction is possible and leads to loss of truncated data.
+"""
+function set_extent_dims(dspace::Dataspace, size::Dims, max_dims::Union{Dims,Nothing} = nothing)
+    checkvalid(obj)
+    rank = length(size)
+    current_size = hsize_t[reverse(new_dims)...]
+    maximum_size = isnothing(max_dims) ? C_NULL : hsize_t[reverse(max_dims)...]
+    h5s_set_extent_simple(obj, rank, current_size, maximum_size)
+    close(obj)
+    return nothing
+end
+
+
+"""
+    HDF5.get_extent_dims(obj::Union{HDF5.Dataspace, HDF5.Dataset, HDF5.Attribute}) -> dims, maxdims
+
+Get the array dimensions from a dataspace, dataset, or attribute and return a tuple of `dims` and `maxdims`.
+"""
+function get_extent_dims(obj::Union{Dataspace,Dataset,Attribute})
+    dspace = obj isa Dataspace ? checkvalid(obj) : dataspace(obj)
+    h5_dims, h5_maxdims = h5s_get_simple_extent_dims(dspace)
+    # reverse dimensions since hdf5 uses C-style order
+    N = length(h5_dims)
+    dims = ntuple(i -> @inbounds(Int(h5_dims[N-i+1])), N)
+    maxdims = ntuple(i -> @inbounds(h5_maxdims[N-i+1]) % Int, N) # allows max_dims to be specified as -1 without triggering an overflow
+    obj isa Dataspace || close(dspace)

```

Comment:
I don't actually understand this line. From L39 `obj` will always be a `Dataspace`, so this check is always true ` close(dspace)` is never called.

Reply:
`obj` can be a `Dataset` or `Attribute`, in which case the dataspace `dspace` is newly opened. It's only the case when `obj` is a literal `Dataspace` that we don't want to close it (since the caller "owns" it).

Reply:
yikes I thought this line was  `dspace isa Dataspace`....